generators/python/src/fern_python/generators/sdk/client_generator/oauth_token_provider_generator.py

```diff
@@ -0,0 +1,489 @@
+                if param.constructor_parameter_name == self._get_client_wrapper_constructor_parameter_name():
+                    continue
+                if param.private_member_name is not None:
+                    writer.write_line(f"self.{param.private_member_name} = {param.constructor_parameter_name}")
+
+            for expr in auth_client_initialization_exprs:
+                writer.write_node(expr)
+
+        return _write_constructor_body
+    
+
+    def _get_auth_client_initialization_exprs(self, client_credentials: ir_types.OAuthClientCredentials) -> List[AST.Expression]:
+        """
+        Generates the initialization of the clients for the token and refresh endpoints. In most cases
+        this is only a single client.
+
+        self._auth_client = AuthClient(client_wrapper=client_wrapper)
+        self._refresh_client = RefreshClient(client_wrapper=client_wrapper)
+        """
+        result: List[AST.Expression] = []
+        
+        token_subpackage_id = self._get_subpackage_id_for_endpoint_id(
+            endpoint_id=client_credentials.token_endpoint.endpoint_reference,
+        )
+
+        result.append(
+            AST.Expression(
+                AST.CodeWriter(
+                    self._get_write_auth_client_initialization(
+                        subpackage_id=token_subpackage_id,
+                        auth_client_member_name=self._get_auth_client_member_name(),
+                    ),
+                ),
+            ),
+        )
+
+        refresh_subpackage_id = self._get_subpackage_id_for_endpoint_id(
+            endpoint_id=client_credentials.refresh_endpoint.endpoint_reference
+        ) if client_credentials.refresh_endpoint is not None else None
+
+        if refresh_subpackage_id is not None and refresh_subpackage_id != token_subpackage_id:
+            self._set_refresh_client_member_name()
+            result.append(
+                AST.Expression(
+                    AST.CodeWriter(
+                        self._get_write_auth_client_initialization(
+                            subpackage_id=refresh_subpackage_id,
+                            auth_client_member_name=self._get_refresh_client_member_name(),
+                        ),
+                    ),
+                ),
+            )
+
+        return result
+    
+    def _get_write_auth_client_initialization(
+            self,
+            subpackage_id: ir_types.SubpackageId,
+            auth_client_member_name: str,
+        ) -> AST.CodeWriterFunction:
+        def _write_auth_client_initialization(writer: AST.NodeWriter) -> None:
+            writer.write_node(AST.Expression(f"self.{auth_client_member_name} = "))
+            writer.write_node(
+                AST.ClassInstantiation(
+                    class_=self._context.get_reference_to_subpackage_service(subpackage_id),
+                    kwargs=[
+                        (
+                            "client_wrapper",
```

Comment:
Great catch here - I'll mark this as a `TODO` for now. 

Although controversial, this might actually be something we restrict in `fern check` to start (i.e. force the auth endpoints to be defined in a nested client). I know we can imagine edge cases that might suggest otherwise, but worth considering given that there are workarounds. Anyway, we can discuss later.